### PR TITLE
Add attendance column to class model

### PR DESCRIPTION
**Pull request by danielchen115**
Classes have a JSON column that tracks attendance. Includes validator and its tests

### CODE REVIEW

File: enrolments/test_validators.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+
+class ValidatorsTestCase(TestCase):
+    def setUp(self):
+        Student.objects.bulk_create(
+            [
+                Student(
+                    first_name="Brittany", last_name="Buckets", attendee_type="Parent"
+                ),
+                Student(first_name="Buckets", last_name="Jr", attendee_type="Child"),
```
Comment (brittanylau) [inline enrolments/test_validators.py]: my fate has been decided

File: enrolments/models.py [hunk 2]
```diff
@@ -13,6 +14,7 @@ class Session(models.Model):
 
     season = models.CharField(max_length=6, choices=SEASON_CHOICES)
     year = models.PositiveSmallIntegerField()
+    attendance = models.JSONField(validators=[validate_attendance])
```
Comment (brittanylau) [inline enrolments/models.py]: should this be on `Class` instead of `Session`?

File: enrolments/validators.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        except ValueError:
+            raise ValidationError(
+                "%(date) must be formatted as YYYY-MM-DD and must be a valid date",
+                code="invalid",
+                params={"date": date},
+            )
+        if attendees and not Student.objects.filter(pk__in=attendees).exists():
+            raise ValidationError(
+                "one or more %(attendee)(s) do not exist",
+                code="invalid",
```
Comment (brittanylau) [inline enrolments/validators.py]: ahh I see what you mean by only the generic "Enter a valid JSON." message showing now

I think the "invalid" code is protected somehow? it seems renaming `code="invalid"` to something like `code="invalid_attendee"` gets your custom message to show!

Screen Shot 2021-03-01 at 1 22 34 AM

File: enrolments/validators.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+            "attendees": [
+                // student IDs
+                1, 2, 3, 4, 5
+            ]
+        }
+    ]
+    """
+    for session in classObj:
+        date = session["date"]
+        attendees = session["attendees"]
```
Comment (brittanylau) [inline enrolments/validators.py:33]: the admin throws a `KeyError` if users don't specify either key in their JSON - can we catch this and raise a validation error?

Screen Shot 2021-03-01 at 1 11 56 AM

File: enrolments/validators.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        try:
+            datetime.strptime(date, "%Y-%m-%d").date()
+        except ValueError:
+            raise ValidationError(
+                "%(date) must be formatted as YYYY-MM-DD and must be a valid date",
+                code="invalid",
+                params={"date": date},
+            )
+        if attendees and not Student.objects.filter(pk__in=attendees).exists():
```
Comment (brittanylau) [inline enrolments/validators.py]: i believe this validates if at least one ID matches a student, even if the rest don't exist – should we instead raise an error if _any_ ID is invalid?

File: enrolments/test_validators.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
+from django.test import TestCase
```
Comment (brittanylau) [inline enrolments/test_validators.py]: nit: can we move this file to `enrolments/tests/test_validators.py` 😄

File: enrolments/models.py [hunk 7]
```diff
@@ -33,6 +34,11 @@ class Class(models.Model):
         blank=True,
     )
 
+    attendance = models.JSONField(validators=[validate_attendance])
+
+    class Meta:
+        verbose_name_plural = "classes"
```
Comment (annoyingpr) [inline enrolments/models.py:41]: What is `verbose_name_plural` used for?

File: enrolments/validators.py [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+        ) == set(attendees):
+            raise ValidationError(
+                "one or more of the following attendee IDs do not exist: "
+                + str(attendees),
+                code="invalid_attendee",
+            )
+
+
+def validate_schema(json, schema, strict=False):
+    """
```
Comment (annoyingpr) [inline enrolments/validators.py:52]: Is there a doc string template we should be using across the board? I'm not sure if there's a standardized one.

**Review by annoyingpr — APPROVED** [review]
🚀  LGTM - just had a few questions for myself

**Review by brittanylau — APPROVED** [review]
this is awesome!!! will be so useful for our other json fields 🚢

File: enrolments/validators.py (hunk 8 above)
Comment (danielchen115) [reply enrolments/validators.py:52]: Do you mean for the schema? The one included in the comments works! I can write out a better-fleshed doc if that helps

File: enrolments/models.py (hunk 7 above)
Comment (danielchen115) [reply enrolments/models.py:41]: It tells Django that the plural of "class" is "classes" instead of "classs". Django by default just adds an s to the end.

**Pull request merged**